### ISSUE TITLE
Post / Put request - Send files (for example: .rfa)

### ISSUE THREAD

**Issue by RitaAguiar** [enmerk4r/Swiftlet#20]
Hi! I was wondering if it's possible to send files (for example: .rfa) with Swiftlet through Post / Put request.
If it is, could you give me guidance on how to do so? Best.

**Comment by enmerk4r** [comment]
Hi @RitaAguiar! Sure. If we are talking about binary files, then take a look at the **Create Byte Array Body** and **Create Byte Array Body From File** components, as well as byte-array-related components in the Utilities section. Essentially, all you need to do is turn your binary file into a byte array using those components and then create a byte array POST body. And if the files are ASCII, then the regular **Create Text Body** component should suffice. Let me know if this works for you.

![image](https://github.com/enmerk4r/Swiftlet/assets/25551621/aec388cf-9404-4d4d-9877-628c92672f00)

**Comment by RitaAguiar** [comment]
Thank you, this is what I was looking for :)
Do you also have a feature to download a file from a link (example: .rfa) as to be used in one of these nodes **Create Byte Array Body** and **Create Byte Array Body From File** ?

**Comment by enmerk4r** [comment]
In terms of getting the file from the server, it's similar. All HTTP request components (such as the GET component, for example) have a Response output. You can use the **Deconstruct Http Response** component in order to get the server response as a byte array and then use the **Byte Array to File** component in order to write it to disk.

![image](https://github.com/enmerk4r/Swiftlet/assets/25551621/427a06d6-7503-4dfa-a3a6-6276537fbe34)

In terms of the Content-Type issue, I think you just found a bug in Swiftlet :) Basically, the ```Content-Type``` header is set on the Body, not on the request itself. As far as I understand, this is a quirk of the HTTP library in C# that I am using because in the HTTP protocol itself, only requests have headers: there is no such thing as a "body header". But in any case, the ```System.Net.Http``` library makes you set the content type header on the Body of the request for whatever reason, which is then supposed to trickle down to the request level. And in Swiftlet's case - it doesn't, because there is a manual step I've missed. In any case, good catch! Thank you. Putting this on the "To Do" list for ```0.1.9```.

**Comment by RitaAguiar** [comment]
Thanks for the help @enmerk4r. I am now able to save files to my local machine using the **Deconstruct Http Response** .

I regards to the post/put of a file, I'm not sure I understood correctly. So I should submit the `ContentType` in both the Body Headers and the headers? If I understood that correctly, I have indeed also tried that, but obtained the same response.

![image](https://github.com/enmerk4r/Swiftlet/assets/44228423/90274626-78a9-4b87-a43c-80afdab79e75)

**Comment by enmerk4r** [comment]
So, the reason why passing `Content-Type` as a request header had no effect is that you forgot the `-` and just passed it as `ContentType` :) However, correcting the typo will not solve your issue, since passing `Content-Type` directly to the POST or PUT component will actually throw an error. The error is due to the HTTP library in .Net requiring Content-Type to be specified on the request body and not on the request itself. That's why the **Create Byte Array Body from File** component has the Content-Type input. 

So if everything worked as intended, you would just specify your type on the Create Body component, and then the POST or PUT component would read it from the body and add it to the request headers. But... there is currently a bug in Swiftlet, where POST and PUT components simply ignore the Content-Type of the body. I will fix this in the next release, but it will obviously take some time (I have a laundry list of things that I am adding and fixing in ```0.1.9```). So for now, I would just use a scripting component in Grasshopper to send your file.

**Comment by RitaAguiar** [comment]
Okay, got it 👍 Can you give me an estimate of when the `0.1.9` version will be released? Thanks!

**Comment by enmerk4r** [comment]
![image](https://github.com/enmerk4r/Swiftlet/assets/25551621/0efc8551-2b30-4fd2-8b28-c6e206e31d4d)


@RitaAguiar I patched the Content-Type issue in the [dev-0-1-9](https://github.com/enmerk4r/Swiftlet/tree/dev-0-1-9) branch. You can either build it from source yourself, or get the updated build here:

[0.1.8-patched-content-type.zip](https://github.com/enmerk4r/Swiftlet/files/11869535/0.1.8-patched-content-type.zip)

If you installed Swiftlet through Yak, you'd have to uninstall your current version through Package Manager before dropping the gha and the dlls into the `Grasshopper\Libraries` folder, otherwise, you will have conflicts.

In terms of the official release, hard to say. I'm working on Swiftlet in my spare time, so this will really depend on when I have a free weekend :) But hopefully soon. In any case, let me know if the ZIP above works for you.

**Comment by RitaAguiar** [comment]
Thanks! I have first uninstalled the previous version and then added the updated build to the `C:\Users\Rita\AppData\Roaming\McNeel\Rhinoceros\packages\7.0\Swiftlet\0.1.8` folder.
I can see that in `_PackageManager` the plug-in is successfully recognized, but when I open Grasshopper I cannot find the components.
Maybe I'm missing something here?

![image](https://github.com/enmerk4r/Swiftlet/assets/44228423/7217c36a-e210-43da-a91c-b6a4ef1eac99)
![image](https://github.com/enmerk4r/Swiftlet/assets/44228423/52bfd8ab-9103-4a58-8ac4-10f7bb92a674)
![image](https://github.com/enmerk4r/Swiftlet/assets/44228423/12605a48-9d02-4921-9cb9-461f7b8a3070)
![image](https://github.com/enmerk4r/Swiftlet/assets/44228423/612b0690-d812-4040-a4f0-537b6594a279)
![image](https://github.com/enmerk4r/Swiftlet/assets/44228423/863f3d62-2c8c-4ae0-aeaf-f6f1f9676e7e)
![image](https://github.com/enmerk4r/Swiftlet/assets/44228423/23f1f410-25d0-4d2f-a2e8-92abf6c2b840)

**Comment by enmerk4r** [comment]
@RitaAguiar No no, don't use Yak for this. Here's what you'll have to do:
1. Uninstall whatever current version of Swiftlet you have on your system
2. Download the [ZIP file](https://github.com/enmerk4r/Swiftlet/files/11869535/0.1.8-patched-content-type.zip)
3. **IMPORTANT:** Make sure to right-click on the downloaded ZIP, go to **Properties**, check the **Unblock** option, and hit **Apply**
4. Unzip the package and place the GHA + DLLs into Grasshopper's `Components Folder` (`AppData/Roaming/Grasshopper/Libraries`). You can make a `Swiftlet` subfolder inside of `Libraries` to keep the installation files together, which will make it easier to uninstall in the future.

In other words, use the ZIP package to install Swiftlet the old-school way :)
Let me know if this works.

**Comment by enmerk4r** [comment]
@RitaAguiar On second thought, you probably could do it your way (replacing the files inside of `McNeel/Rhinoceros/packages/...`), but I wouldn't touch the manifest. You can just replace the GHA and the DLL files in your 0.1.8 installation. This will make it easier to upgrade to 0.1.9 later. In this case, **unblocking the ZIP file** before unzipping is still a crucial step.

**Comment by RitaAguiar** [comment]
I am now able to use the patched version. :)
I have removed the `ContentType` input from the headers as it was not necessary. Now it is only being provided through the body of the request. The `multipart/form-data` seems to be the only format that is allowed as I also tried with the other different file formats but it says they are not supported. However, I did not specify any mandatory format in my endpoint. All formats are allowed and I have tested this with a Python Script and I am allowed to upload anything.
I've tried to find a way to provide a boundary as it is being asked here, but I'm not sure about this step. Do you have any suggestions?
Thank you for your help!
![image](https://github.com/enmerk4r/Swiftlet/assets/44228423/48b920b4-a0e2-43a5-96b4-3cd6d0dc01ee)

**Comment by enmerk4r** [comment]
@RitaAguiar ok, I'm glad the `Content-Type` is working now! With regard to the multipart data, it's difficult to diagnose the issue without seeing the server-side code. Would you be willing to share at least a toy "hello world" example that can help reproduce the error?

**Comment by RitaAguiar** [comment]
When trying with `application/octet-stream` the `Material update failed` exception is thrown.
I can change this endpoint however I need so that this GHscript call works.

Endpoint definition:

```
class updateMaterialBIMModel(APIView):
    def post(self, request, *args, **kwargs):
        if not request.headers["materialID"]:
            return Response("The 'materialID' field is required", status=400)

        try:
            material = Material.objects.get(id=int(request.headers["materialID"]))
        except Material.DoesNotExist:
            return Response("Material ID does not exist", status=404)

        try:
            bim_model_file = request.FILES.get("BIMModel") # this is where the code is failing
        except Exception as e:
            return Response(f"Material update failed: {e}", status=e.status_code)

        if not bim_model_file:
            return Response("The 'BIMModel' file is required", status=400)

        # Read the contents of the file into a byte array
        try:
            bim_model_bytes = bim_model_file.read()
        except Exception as e:
            return Response(f"Failed to read 'BIMModel' file: {str(e)}", status=400)

        # Encode the byte array as base64
        try:
            bim_model_data = base64.b64encode(bim_model_bytes).decode("utf-8")
        except Exception as e:
            return Response(f"Failed to encode 'BIMModel' data: {str(e)}", status=500)

        # Save the byte array data to a file or handle it as needed
        # For example, save it to a file using a unique name
        file_name = f"bim_model_{material.id}.rfa"
        file_path = os.path.join("materials/bim_models", file_name)

        try:
            with open(file_path, "wb") as file:
                file.write(bim_model_bytes)
        except Exception as e:
            return Response(f"Failed to save 'BIMModel' data: {str(e)}", status=500)

        # Update the material's bim_model field
        material.bim_model.name = file_path
        material.save()

        return Response(f"Material updated: {file_name}", status=200)
```

Working Python script call:
```
def update_material_bim_model(material_id, bim_model):
    url = f"{NEXT_PUBLIC_API_URL}/materials/update-material-bim-model/"
    headers = {
        "materialID": str(material_id),
    }
    files = {
        "BIMModel": open(bim_model, "rb"),
    }
    response = requests.post(url, headers=headers, files=files)
    if response.status_code == 200:
        data = response.json()
        # Process the response data as needed
        print(data)
        return data
    else:
        print("Error:", response.status_code)
        print("response.text", response.text)
```

**Comment by enmerk4r** [comment]
@RitaAguiar Alright, I guess you learn something new every day haha. Basically, the **Byte Array Body From File** component uses a `ByteArrayContent` type in C# in order to describe the `HttpContent`. That just won't work for an endpoint that requires a multipart form, because what we are giving it is not a "form", but a simple stream of binary data. So sounds like I need to add a new component that uses `MultipartFormDataContent` under the hood and actually lets you "fill out" various fields of the form. Give me a few days to test this out, I can send you another build once it's ready.

If you need a quicker fix, then just don't use MultipartForm in your endpoint - you don't need it to get the data. The **Byte Array Body From File** component will send your entire file as a byte array in the body of a request. You can then just read that byte array from `request.body` and save it to a file or do whatever else you need with it. All of the additional metadata (such as the file name, for example) can be provided in the request headers or in the query params.

If you decide to go the byte array route, I would also recommend using the GZIP component in Swiftlet to compress your file before sending it. You can then un-GZIP it on the endpoint side. 

In any case, I'll try to add the MultipartData components in the next few days, but otherwise - you have options :)

**Comment by RitaAguiar** [comment]
Thanks @enmerk4r ! I went with the quicker fix option with Byte Array. This workflow worked.
Then I tried adding the compressing and decompressing. However, I am not able to correctly decompress the files, neither .rfa.gz or .png.gz. I tried with the gzip, the zlib and the zipfile libraries in python. In the case I provide a .png.gz I get `Failed to save 'image' data: 'utf-8' codec can't decode byte 0x8b in position 1: invalid start byte`. In the case I provide a .rfa.gz I get `Unexpected character encountered while parsing value: <. Path '', line 0, position 0.`
![image](https://github.com/enmerk4r/Swiftlet/assets/44228423/2f4dd4da-2aca-4859-b7d1-e2600892979d)
But yeah, without compression this works great.

**Comment by enmerk4r** [comment]
@RitaAguiar Ok! Great to hear! And just out of curiosity - are you trying to un-GZIP the binary data inside the route before you write it to disk? Or are you writing it to disk first and then trying to open it? As far as I know, the latter likely won't work. I would recommend decompressing the binary data that comes from the POST request before doing anything else to it. But in any case, glad it's all working now.

**Comment by RitaAguiar** [comment]
I've been trying with first read, then decompress, then write:
```
with open(image_file, "rb") as file_in:
    compressed_data = file_in.read()
decompressed_data = gzip.decompress(compressed_data)
with open(file_path, "wb") as file_out:
    file_out.write(decompressed_data)
```

**Comment by enmerk4r** [comment]
@RitaAguiar Hey! Idk if you still need help with `multipart/form-data`, but I just pushed out `0.1.9` to both Yak and Food4Rhino. Swiftlet now has two components to deal with Multipart data: **Create Multipart Form Body Named** and **Create Multipart Form Body Unnamed** (the component names are getting out of hand at this point hehe). Both components take a list of other **HTTP Body** types that Swiftlet already supports (so either a text-based body or a binary body). The **Named** component lets you also pass a list of field names, which essentially turns your form into a key-value dictionary. The **Unnamed** component just combines all of the bodies you pass to it into a list. So, your definition would look something like this:

![image](https://github.com/enmerk4r/Swiftlet/assets/25551621/ac49ba2f-c0dd-4730-b235-03f14d698bbc)

I know we found a different workaround for your use case, but I feel like it's nice to have support for multipart data in Swiftlet. Thanks for pointing that out!
